### PR TITLE
Affine transforms

### PR DESCRIPTION
**Pull request by andyferris**
We implement some abstract and concrete types of Affine and linear transformations (as well as translations). A quick overview: 

* A `LinearTransformation` is more-or-less a wrapper for a matrix or matrix-like object which can be chained with `compose` and called to result in matrix multiplication.
* A `Translation` is a similar wrapper for adding a vector or vector-like object.
* An `AffineTransformation` has both of these - first multiplying by the matrix, then adding the vector. This is our canonical choice, but it is well documented.
* These compose together in sensible ways, e.g. by turning two Affine transformations into just one.

There are also related functions such as `translation_vector()` and `transformation_matrix()` for extracting the parameters from an `AffineTransformation`, etc. If you need to interact with code which uses the other convention for Affine transformations (first add the vector, and then multiply by a matrix), you can get that vector with `translation_vector_reverse()`.

A note on rotations: I have a plan to allow the objects in `Rotations.jl` to be more matrix-like and simply remove the rotations here sometime in the future - they will just be a `LinearTransformation` with a different kind of matrix. This removes duplication.

Final request for reviewers: we can discuss shorted forms for `LinearTransformation` and `AffineTransformation`. While these names are quite clear (on purpose), the *CoordinateTransformations* package tends to be a bit wordy. Options like `LinearTrans` or `LinearMap` may be considered.

### CODE REVIEW

**Comment by c42f** [comment]
Hmm, something odd going on with the travis build here, I restarted it.

**Comment by c42f** [comment]
Right, build looks ok now.  0.5 still broken as expected due to Rotations.jl

File: src/CoordinateTransformations.jl [hunk 1]
```diff
@@ -40,15 +40,21 @@ export SphericalFromCartesian, CartesianFromSpherical,
        CylindricalFromSpherical, SphericalFromCylindrical
 
 # Common transformations
-export Translation
+export AbstractAffineTransformation
+export AffineTransformation, LinearTransformation, Translation, transformation_matrix, translation_vector, translation_vector_reverse
+export affine_decomposition_T_of_L, affine_decomposition_L_of_T
```
Comment (c42f) [inline src/CoordinateTransformations.jl]: Oops, these need to go.

File: src/CoordinateTransformations.jl [hunk 2]
```diff
@@ -40,15 +40,21 @@ export SphericalFromCartesian, CartesianFromSpherical,
        CylindricalFromSpherical, SphericalFromCylindrical
 
 # Common transformations
-export Translation
+export AbstractAffineTransformation
```
Comment (c42f) [inline src/CoordinateTransformations.jl]: Should we export AbstractLinearTransformation?

File: src/commontransformations.jl [hunk 3]
```diff
@@ -1,41 +1,283 @@
...
+# Could try do similar for transform_deriv_params()?
+
+transform_deriv(trans::AbstractAffineTransformation, x) = transformation_matrix(trans)
+
+function Base.inv(trans::AbstractAffineTransformation)
+    Minv = inv(transformation_matrix(trans))
+    AffineTransformation(Minv, -Minv * translation_vector(trans))
+end
+
+function Base.isapprox(t1::AbstractAffineTransformation, t2::AbstractAffineTransformation, kwargs...)
```
Comment (c42f) [inline src/commontransformations.jl]: should have a semicolon before the `kwargs`

File: src/commontransformations.jl [hunk 4]
```diff
@@ -1,41 +1,283 @@
...
+end
+
+# Default implementations
+@compat function (trans::AbstractLinearTransformation)(x)
+    transformation_matrix(trans) * x
+end
+
+# transform_deriv() identical to that provided by AbstractAffineTransformation
+
+function Base.isapprox(t1::AbstractLinearTransformation, t2::AbstractLinearTransformation, kwargs...)
```
Comment (c42f) [inline src/commontransformations.jl]: Another, `kwargs` issue, ditto below

File: src/commontransformations.jl [hunk 5]
```diff
@@ -1,41 +1,283 @@
...
+Specific implementations may provide equivalent specializations of `call`, etc,
+for optimization purposes. The major difference with AbstractLinearTransformation
+is that the matrix is assumed to be orthogonal/unitary for the purpose of
+inverting the transformation.
+
+(See also Rotation, AxisRotation, AbstractLinearTransformation)
+"""
+abstract AbstractRotation <: AbstractLinearTransformation
+
+transformation_matrix(trans::AbstractRotation) = error("AbstractRotation $(typeof(trans)) must implement transformation_matrix()")
```
Comment (c42f) [inline src/commontransformations.jl]: Is this part of the future stuff you want to clean up rotations with?

File: src/commontransformations.jl [hunk 6]
```diff
@@ -85,7 +327,12 @@ end
     Polar(x.r, x.θ + trans.angle)
 end
 
-function transform_deriv(trans::Rotation2D, x)
+function transformation_matrix(trans::Rotation2D)
+    @fsa [ trans.cos -trans.sin;
+           trans.sin  trans.cos ]
+end
```
Comment (c42f) [inline src/commontransformations.jl]: Repeated below?

File: src/commontransformations.jl [hunk 7]
```diff
@@ -375,20 +624,43 @@ function transform_deriv(trans::RotationXY, x)
            trans.sin  trans.cos Z;
            Z          Z         I]
 end
```
Comment (c42f) [inline src/commontransformations.jl]: Can we just delete `transform_deriv`  for these rotations now?

File: test/commontransformations.jl [hunk 8]
```diff
@@ -333,11 +345,11 @@
             trans = euler_rotation(0.1,0.2,0.3)
             x2 = Point(2.730537054338937,0.8047190852558106,2.428290466296628)
 
-            @test trans.t1.t1 == RotationXY(0.1)
-            @test trans.t1.t2 == RotationYZ(0.2)
-            @test trans.t2 == RotationZX(0.3)
+            #@test trans.t1.t1 == RotationXY(0.1)
+            #@test trans.t1.t2 == RotationYZ(0.2)
+            #@test trans.t2 == RotationZX(0.3)
```
Comment (c42f) [inline test/commontransformations.jl]: Are these gone on purpose?  Should they be deleted?

File: test/commontransformations.jl [hunk 9]
```diff
@@ -353,14 +365,14 @@
...
-            m = transform_deriv_params(trans, x)
-            @test m ≈ m_gn
+            #trans_gn = euler_rotation(Dual(0.1, (1.0, 0.0, 0.0)), Dual(0.2, (0.0, 1.0, 0.0)), Dual(0.3, (0.0, 0.0, 1.0)))
+            #x = Point(2.0,1.0,3.0)
+            #x2_gn = trans_gn(x)
+            #m_gn = @fsa [partials(x2_gn[1], 1) partials(x2_gn[1], 2) partials(x2_gn[1], 3);
+            #             partials(x2_gn[2], 1) partials(x2_gn[2], 2) partials(x2_gn[2], 3);
+            #             partials(x2_gn[3], 1) partials(x2_gn[3], 2) partials(x2_gn[3], 3)]
+            #m = transform_deriv_params(trans, x)
+            #@test m ≈ m_gn
```
Comment (c42f) [inline test/commontransformations.jl]: Intentional?

File: src/CoordinateTransformations.jl (hunk 2 above)
Comment (c42f) [inline src/CoordinateTransformations.jl]: And what about AbstractTranslation?

File: src/commontransformations.jl [hunk 10]
```diff
@@ -1,41 +1,283 @@
...
+(See also LinearTransformation, AbstractAffineTransformation)
+"""
+abstract AbstractLinearTransformation <: AbstractAffineTransformation
+
+transformation_matrix(trans::AbstractLinearTransformation) = error("AbstractLinearTransformation $(typeof(trans)) must implement transformation_matrix()")
+function translation_vector(trans::AbstractLinearTransformation)
+    m = transformation_matrix(trans)
+    s = size(m, 1)
+    T = eltype(m)
+    return zeros(T, s)
```
Comment (c42f) [inline src/commontransformations.jl]: This isn't ideal for use with FixedSizeArrays.  Maybe it doesn't matter because calling `translation_vector` on a LinearTransformation isn't really something that makes a whole lot of sense?

**Comment by c42f** [comment]
@timholy here's a go at some reasonably generic affine transformation types which should basically work for both fixed size and normal arrays.  It's pretty close to done I think, but feedback is very welcome.

**Comment by c42f** [comment]
@andyferris I like the generalization you've got here.

Some of the type names are indeed a bit of a mouthful.  `AffineMap` and `LinearMap` are definitely a lot nicer to write, just not so consistent...

File: src/commontransformations.jl (hunk 10 above)
Comment (andyferris) [inline src/commontransformations.jl]: Right. I guess we can check if `m` is isbits and return the appropriate type.

File: src/commontransformations.jl (hunk 3 above)
Comment (andyferris) [inline src/commontransformations.jl]: Thanks

File: src/commontransformations.jl (hunk 5 above)
Comment (andyferris) [inline src/commontransformations.jl]: Sorry I thought I pushed the deletion of this

File: test/commontransformations.jl (hunk 8 above)
Comment (andyferris) [inline test/commontransformations.jl]: I'll clean all this up when rotations are changed. For now it is a reminder of the old behaviour while I consider how we might want axis rotations to compose into Euler angles (or just into a matrix).

File: test/commontransformations.jl (hunk 9 above)
Comment (andyferris) [inline test/commontransformations.jl]: This will also be sorted out later.

File: src/commontransformations.jl (hunk 6 above)
Comment (andyferris) [inline src/commontransformations.jl]: Thanks

**Comment by andyferris** [comment]
OK I've addressed the code review.

**Comment by andyferris** [comment]
> @andyferris I like the generalization you've got here.

Thanks! Though they are only a continuation on your ideas.

> Some of the type names are indeed a bit of a mouthful. AffineMap and LinearMap are definitely a lot nicer to write, just not so consistent...

Right. I'm thinking for consistency we stick with `*Transformation` until there is a compelling reason/alternative, unless there is strong opposition?

**Comment by c42f** [comment]
> I'm thinking for consistency we stick with *Transformation

Works for me.

File: test/commontransformations.jl (hunk 8 above)
Comment (c42f) [inline test/commontransformations.jl]: Sure works for me.  Can be worth adding a TODO to make it clear why a code block was commented out otherwise it sometimes ends up hanging around.

**Comment by c42f** [comment]
I'm good with this.

**Comment by andyferris** [comment]
OK I'll merge
